server/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -829,7 +831,12 @@ public void executePipeline(Pipeline pipeline, BiConsumer<IngestDocument, Except
             return;
         }
 
-        if (executedPipelines.add(pipeline.getId())) {
+        if (executedPipelines.size() >= MAX_PIPELINES) {
```

Comment:
I'm just thinking aloud to myself in a comment: if we have already executed X pipelines, and X is the max, then we should not execute X+1 pipelines because then we'd be executing more pipelines than the max. So there's not an off by one error here, and I'm in agreement with this logic. For the record I was considering if this if was misplaced and might have been better after the `executedPipelines.add(...)` call below, but this is right and good.